File: electron/server/parents.js
// Nickland Edusoft — Parent identity for the mobile API
// Copyright © 2026 Nickland Sales. All rights reserved.
//
// Parents are NOT staff users. A parent account links a phone/email + password
// to one or more students, verified against the guardian contact/email fields
// already stored on each student. Passwords are hashed with scrypt (built-in
// Node crypto — no external dependency, so this layer is fully self-contained
// and testable).

const crypto = require('crypto');
const { postToOutbox, syncEnabled } = require('./sync/outbox');

function hashPassword(password) {
  const salt = crypto.randomBytes(16).toString('hex');
  const derived = crypto.scryptSync(String(password), salt, 64).toString('hex');
  return `scrypt$${salt}$${derived}`;
}
function verifyPassword(password, stored) {
  if (!stored || !stored.startsWith('scrypt$')) return false;
  const [, salt, expected] = stored.split('$');
  const derived = crypto.scryptSync(String(password), salt, 64).toString('hex');
  const a = Buffer.from(derived, 'hex'); const b = Buffer.from(expected, 'hex');
  return a.length === b.length && crypto.timingSafeEqual(a, b);
}

function normPhone(raw) {
  let s = String(raw || '').replace(/[^\d+]/g, '');
  if (!s) return '';
  if (s.startsWith('+')) s = s.slice(1);
  if (s.startsWith('00')) s = s.slice(2);
  if (s.startsWith('0')) s = '233' + s.slice(1);
  else if (s.length === 9) s = '233' + s;
  return s;
}

// Find students whose guardian contact/email matches the given phone/email.
function matchStudentsByContact(db, { phone, email }) {
  const results = new Map();
  if (phone) {
    const np = normPhone(phone);
    const rows = db.prepare(`
      SELECT id, father_contact, mother_contact, guardian_contact,
             father_name, mother_name, guardian_name
      FROM students WHERE status = 'Active'
        AND (father_contact IS NOT NULL OR mother_contact IS NOT NULL OR guardian_contact IS NOT NULL)
    `).all();
    for (const s of rows) {
      const rel =
        normPhone(s.father_contact) === np ? 'Father' :
        normPhone(s.mother_contact) === np ? 'Mother' :
        normPhone(s.guardian_contact) === np ? 'Guardian' : null;
      if (rel) results.set(s.id, rel);
    }
  }
  if (email) {
    const e = String(email).trim().toLowerCase();
    const rows = db.prepare(`
      SELECT id, father_email, mother_email, guardian_email FROM students WHERE status = 'Active'
    `).all();
    for (const s of rows) {
      const rel =
        (s.father_email || '').toLowerCase() === e ? 'Father' :
        (s.mother_email || '').toLowerCase() === e ? 'Mother' :
        (s.guardian_email || '').toLowerCase() === e ? 'Guardian' : null;
      if (rel && !results.has(s.id)) results.set(s.id, rel);
    }
  }
  return [...results.entries()].map(([student_id, relationship]) => ({ student_id, relationship }));
}

function linkStudents(db, parentId, matches) {
  const ins = db.prepare('INSERT OR IGNORE INTO parent_students (parent_id, student_id, relationship) VALUES (?, ?, ?)');
  for (const m of matches) ins.run(parentId, m.student_id, m.relationship);
}

// Self-registration: only succeeds if the phone/email matches a student on file.
function registerParent(db, { full_name, phone, email, password }) {
  if (!password || String(password).length < 4) return { ok: false, error: 'Password must be at least 4 characters.' };
  if (!phone && !email) return { ok: false, error: 'Phone or email is required.' };
  const matches = matchStudentsByContact(db, { phone, email });
  if (matches.length === 0) {
    return { ok: false, error: 'No student found with that phone/email. Ask the school to register you.' };
  }
  const np = phone ? normPhone(phone) : null;
  const existing = np ? db.prepare('SELECT id FROM parents WHERE phone = ?').get(np) : null;
  if (existing) return { ok: false, error: 'An account already exists for this phone. Please log in.' };
  const r = db.prepare(`
    INSERT INTO parents (full_name, phone, email, password_hash) VALUES (?, ?, ?, ?)
  `).run(full_name || 'Parent', np, (email || '').trim().toLowerCase() || null, hashPassword(password));
  linkStudents(db, r.lastInsertRowid, matches);
  enqueueParentAuth(db, r.lastInsertRowid);
  return { ok: true, parent_id: r.lastInsertRowid, linked: matches.length };
}

// Admin-provisioned parent with explicit student links (+ optional contact match).
function provisionParent(db, { full_name, phone, email, password, studentIds = [], createdBy }) {
  const np = phone ? normPhone(phone) : null;
  if (np) {
    const dup = db.prepare('SELECT id FROM parents WHERE phone = ?').get(np);
    if (dup) return { ok: false, error: 'A parent with this phone already exists.' };
  }
  const pass = password || Math.random().toString(36).slice(2, 8);
  const r = db.prepare(`
    INSERT INTO parents (full_name, phone, email, password_hash, must_change_password, created_by)
    VALUES (?, ?, ?, ?, ?, ?)
  `).run(full_name || 'Parent', np, (email || '').trim().toLowerCase() || null, hashPassword(pass),
         password ? 0 : 1, createdBy || null);
  const parentId = r.lastInsertRowid;
  const explicit = (studentIds || []).map(id => ({ student_id: id, relationship: 'Guardian' }));
  const matched = matchStudentsByContact(db, { phone, email });
  linkStudents(db, parentId, [...explicit, ...matched]);
  enqueueParentAuth(db, parentId);
  return { ok: true, parent_id: parentId, temp_password: password ? undefined : pass };
}

function loginParent(db, { identifier, password }) {
  const np = normPhone(identifier);
  let parent = np ? db.prepare('SELECT * FROM parents WHERE phone = ?').get(np) : null;
  if (!parent && identifier) {
    parent = db.prepare('SELECT * FROM parents WHERE lower(email) = ?').get(String(identifier).trim().toLowerCase());
  }
  if (!parent || !parent.is_active) return { ok: false, error: 'Invalid credentials.' };
  if (!verifyPassword(password, parent.password_hash)) return { ok: false, error: 'Invalid credentials.' };
  db.prepare("UPDATE parents SET last_login = datetime('now') WHERE id = ?").run(parent.id);
  return { ok: true, parent: { id: parent.id, full_name: parent.full_name, phone: parent.phone, email: parent.email, must_change_password: !!parent.must_change_password } };
}

function studentIdsForParent(db, parentId) {
  return db.prepare('SELECT student_id FROM parent_students WHERE parent_id = ?').all(parentId).map(r => r.student_id);
}

// Project a thin parent auth record to the cloud so the web portal can log the
// parent in even when the desktop is offline. No-op unless cloud sync is on.
// The desktop remains the source of truth for the account.
function enqueueParentAuth(db, parentId) {
  try {
    if (!syncEnabled(db)) return null;
    const p = db.prepare('SELECT id, full_name, phone, email, password_hash, is_active FROM parents WHERE id = ?').get(parentId);
    if (!p) return null;
    const studentKeys = studentIdsForParent(db, parentId).map(id => `student:${id}`);
    return postToOutbox(db, {
      entity_type: 'parent_auth',
      entity_key: `parent:${parentId}`,
      op: p.is_active ? 'upsert' : 'delete',
      payload: {
        parent_id: parentId, full_name: p.full_name, phone: p.phone, email: p.email,
        password_hash: p.password_hash, is_active: p.is_active, student_keys: studentKeys,
      },
    });
  } catch (_) { return null; }
}

function listParents(db) {
  return db.prepare(`
    SELECT p.id, p.full_name, p.phone, p.email, p.is_active, p.last_login, p.created_at,
           (SELECT COUNT(*) FROM parent_students ps WHERE ps.parent_id = p.id) AS child_count
    FROM parents p ORDER BY p.created_at DESC
  `).all();
}

function setParentPassword(db, parentId, newPassword) {
  if (!newPassword || String(newPassword).length < 4) return { ok: false, error: 'Password too short.' };
  db.prepare('UPDATE parents SET password_hash = ?, must_change_password = 0 WHERE id = ?')
    .run(hashPassword(newPassword), parentId);
  enqueueParentAuth(db, parentId);
  return { ok: true };
}

module.exports = {
  hashPassword, verifyPassword, normPhone, matchStudentsByContact,
  registerParent, provisionParent, loginParent, studentIdsForParent,
  listParents, setParentPassword, enqueueParentAuth,
};

File: cloud-python/app/school/parents.py
"""Parents — a translation of ``electron/server/parents.py``'s JavaScript twin.

A parent is not a member of staff and never becomes one. They have their own
table, their own password hashing (scrypt, as the offline system uses, so
nobody has to be re-enrolled), and a session that can reach exactly one thing:
their own children.

The link between a parent and a child is verified against the guardian contact
details already on the pupil's record. That is the whole security model and it
is a good one: a school does not have to remember to link anybody, and somebody
who does not appear on a child's record cannot claim them by typing an id.

Every read below filters by the children this parent actually has, resolved
from the database on the request. There is no endpoint that takes a student id
and trusts it.
"""
import datetime
import hashlib
import hmac
import re
import secrets

from .billing import round2

_NON_DIGIT = re.compile(r"[^\d+]")


def hash_password(password):
    salt = secrets.token_hex(16)
    derived = hashlib.scrypt(str(password).encode(), salt=salt.encode(),
                             n=16384, r=8, p=1, dklen=64).hex()
    return f"scrypt${salt}${derived}"


def verify_password(password, stored):
    if not stored or not str(stored).startswith("scrypt$"):
        return False
    try:
        _, salt, expected = str(stored).split("$")
        derived = hashlib.scrypt(str(password).encode(), salt=salt.encode(),
                                 n=16384, r=8, p=1, dklen=64).hex()
        return hmac.compare_digest(derived, expected)
    except Exception:
        return False


def norm_phone(raw):
    """Ghanaian numbers, however they were typed.

    0244…, +233244…, 00233244… and 244… are one number, and a parent who typed
    it differently at registration should still be able to sign in.
    """
    s = _NON_DIGIT.sub("", str(raw or ""))
    if not s:
        return ""
    if s.startswith("+"):
        s = s[1:]
    if s.startswith("00"):
        s = s[2:]
    if s.startswith("0"):
        s = "233" + s[1:]
    elif len(s) == 9:
        s = "233" + s
    return s


def match_students(db, phone=None, email=None):
    """Which children this contact appears on."""
    found = {}
    if phone:
        np = norm_phone(phone)
        if np:
            for s in db.all("""SELECT id, father_contact, mother_contact, guardian_contact
                                 FROM students WHERE status = 'Active'"""):
                for column, relationship in (("father_contact", "Father"),
                                             ("mother_contact", "Mother"),
                                             ("guardian_contact", "Guardian")):
                    if norm_phone(s[column]) == np:
                        found.setdefault(s["id"], relationship)
                        break
    if email:
        e = str(email).strip().lower()
        if e:
            for s in db.all("""SELECT id, father_email, mother_email, guardian_email
                                 FROM students WHERE status = 'Active'"""):
                for column, relationship in (("father_email", "Father"),
                                             ("mother_email", "Mother"),
                                             ("guardian_email", "Guardian")):
                    if (s[column] or "").lower() == e:
                        found.setdefault(s["id"], relationship)
                        break
    return [{"student_id": k, "relationship": v} for k, v in found.items()]


def register(db, full_name, phone=None, email=None, password=None):
    """Self-registration, which only works if the school already has you.

    A parent whose number is not on any pupil's record is told to ask the
    school. That is deliberate: the alternative is an account that claims a
    child by asserting a relationship nobody checked.
    """
    if not password or len(str(password)) < 8:
        return {"ok": False, "status": 400, "error": "A password must be at least 8 characters."}
    if not phone and not email:
        return {"ok": False, "status": 400, "error": "A phone number or an email is required."}
    if db.get_setting("mobile_parent_self_register", "true") != "true":
        return {"ok": False, "status": 403,
                "error": "Registration is closed. Ask the school to set up your account."}

    matches = match_students(db, phone, email)
    if not matches:
        return {"ok": False, "status": 400,
                "error": "We could not find a pupil with those contact details. Ask the school."}

    np = norm_phone(phone) if phone else None
    if np and db.one("SELECT id FROM parents WHERE phone = %s", (np,)):
        return {"ok": False, "status": 400, "error": "An account already exists for this number."}

    parent_id = db.insert("parents", {
        "full_name": full_name or "Parent", "phone": np,
        "email": (email or "").strip().lower() or None,
        "password_hash": hash_password(password), "is_active": 1,
    })
    for m in matches:
        db.run("""INSERT INTO parent_students (parent_id, student_id, relationship)
                       VALUES (%s,%s,%s) ON CONFLICT DO NOTHING""",
               (parent_id, m["student_id"], m["relationship"]))
    return {"ok": True, "parent_id": parent_id, "linked": len(matches)}


def sign_in(db, identifier, password):
    np = norm_phone(identifier)
    em = str(identifier or "").strip().lower()
    parent = db.one("""SELECT * FROM parents
                        WHERE is_active = 1 AND (phone = %s OR lower(email) = %s)""", (np, em))
    if not parent or not verify_password(password, parent["password_hash"]):
        return {"ok": False, "status": 401,
                "error": "Those details did not match an account. Check and try again."}
    return {"ok": True, "parent": {"id": parent["id"], "full_name": parent["full_name"],
                                   "phone": parent["phone"], "email": parent["email"],
                                   "must_change_password": bool(parent["must_change_password"])}}


def student_ids(db, parent_id):
    return [r["student_id"] for r in
            db.all("SELECT student_id FROM parent_students WHERE parent_id = %s", (parent_id,))]


def owns(db, parent_id, student_id):
    return bool(db.one("""SELECT 1 FROM parent_students
                           WHERE parent_id = %s AND student_id = %s""", (parent_id, student_id)))


def children(db, parent_id):
    """Every child of this parent, with the figures a parent actually opens for."""
    ids = student_ids(db, parent_id)
    if not ids:
        return {"ok": True, "children": []}
    term = db.one("SELECT id, label FROM terms WHERE is_current = 1")
    rows = db.all("""
      SELECT s.id, s.index_number, s.surname, s.first_name, s.other_names, s.gender, s.status,
             c.name AS class_name,
             (SELECT TRIM(COALESCE(st.surname,'') || ' ' || COALESCE(st.first_name,''))
                FROM staff_assignments sa JOIN staff st ON st.id = sa.staff_id
               WHERE sa.class_group_id = c.id AND sa.is_class_teacher = 1 LIMIT 1) AS class_teacher
        FROM students s LEFT JOIN class_groups c ON c.id = s.current_class_id
       WHERE s.id = ANY(%s) ORDER BY s.surname, s.first_name""", (ids,))
    for r in rows:
        r["name"] = f"{r.get('surname') or ''} {r.get('first_name') or ''}".strip()
        bill = db.one("""SELECT total_billed, total_paid, balance FROM student_bills
                          WHERE student_id = %s AND term_id = %s
                            AND COALESCE(status,'active') = 'active'""",
                      (r["id"], term["id"] if term else None)) if term else None
        r["fees"] = {"billed": round2((bill or {}).get("total_billed") or 0),
                     "paid": round2((bill or {}).get("total_paid") or 0),
                     "balance": round2((bill or {}).get("balance") or 0)}
        att = db.one("""SELECT count(*) FILTER (WHERE status = 'present') AS present,
                               count(*) FILTER (WHERE status = 'absent') AS absent
                          FROM student_attendance WHERE student_id = %s AND term_id = %s""",
                     (r["id"], term["id"] if term else None)) if term else None
        r["attendance"] = att or {"present": 0, "absent": 0}
    return {"ok": True, "term": term, "children": rows}


def child(db, parent_id, student_id):
    """One child, in full: the bill line by line, the receipts, the marks."""
    if not owns(db, parent_id, student_id):
        return {"ok": False, "status": 403, "error": "Not your child."}
    term = db.one("SELECT * FROM terms WHERE is_current = 1")
    student = db.one("""
      SELECT s.id, s.index_number, s.surname, s.first_name, s.other_names, s.gender,
             s.date_of_birth, s.status, c.name AS class_name
        FROM students s LEFT JOIN class_groups c ON c.id = s.current_class_id
       WHERE s.id = %s""", (student_id,))
    if not student:
        return {"ok": False, "status": 404, "error": "Not found."}
    student["name"] = f"{student.get('surname') or ''} {student.get('first_name') or ''}".strip()

    bill = db.one("""SELECT * FROM student_bills WHERE student_id = %s AND term_id = %s
                       AND COALESCE(status,'active') = 'active'""",
                  (student_id, term["id"])) if term else None
    return {
        "ok": True, "child": student,
        "term": {"id": term["id"], "label": term["label"]} if term else None,
        "bill": {
            "billed": round2(bill["total_billed"]), "paid": round2(bill["total_paid"]),
            "balance": round2(bill["balance"]), "arrears": round2(bill["arrears_from_prev"]),
            "discount": round2(bill["discount_amount"]),
        } if bill else None,
        "items": db.all("""SELECT item_number, description, amount, is_arrear
                             FROM bill_line_items WHERE student_bill_id = %s
                            ORDER BY is_arrear, item_number, id""", (bill["id"],)) if bill else [],
        "payments": db.all("""
          SELECT receipt_number, amount, payment_date, payment_method, is_reversed
            FROM payments WHERE student_id = %s AND is_reversed = 0
           ORDER BY payment_date DESC, id DESC LIMIT 60""", (student_id,)),
        "results": db.all("""
          SELECT sub.name AS subject, sc.class_score, sc.exam_score, sc.total_score, sc.grade_remark
            FROM scores sc JOIN subjects sub ON sub.id = sc.subject_id
           WHERE sc.student_id = %s AND sc.term_id = %s ORDER BY sub.name""",
                          (student_id, term["id"])) if term else [],
        "conduct": db.all("""SELECT event_type, title, description, date FROM student_events
                              WHERE student_id = %s ORDER BY date DESC, id DESC LIMIT 20""",
                          (student_id,)),
        "attendance": db.all("""SELECT date, status FROM student_attendance
                                 WHERE student_id = %s AND term_id = %s ORDER BY date DESC LIMIT 90""",
                             (student_id, term["id"])) if term else [],
        "homework": db.all("""
          SELECT h.title, h.description, h.due_date, s.name AS subject_name,
                 hs.status, hs.marks, h.max_marks
            FROM homework h
            LEFT JOIN subjects s ON s.id = h.subject_id
            LEFT JOIN homework_submissions hs ON hs.homework_id = h.id AND hs.student_id = %s
           WHERE h.class_group_id = (SELECT current_class_id FROM students WHERE id = %s)
           ORDER BY h.due_date DESC NULLS LAST LIMIT 30""", (student_id, student_id)),
    }


def announcements(db, parent_id):
    mine = student_ids(db, parent_id)
    return {"ok": True, "announcements": db.all("""
      SELECT id, title, body, audience, target_student_id, created_at
        FROM announcements
       WHERE is_active = 1 AND (audience = 'all'
             OR (audience = 'student' AND target_student_id = ANY(%s)))
       ORDER BY created_at DESC LIMIT 50""", (mine or [0],))}
